WFLY-3893 JacORB compatibility subsystem

tadamski opened the pull request:
JacORB compatibility subsystem uses OpenJDK ORB to run jacorb subsystem configuration. Configuration parameters are gathered, transformed and, if possible, use to start an ORB. If attribute cannot be emulated then subsystem rejects it and doesn't start.

tadamski committed:
WFLY-3893 JacORB compatibility subsystem

ctomc commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBSubsystemAdd.java:
```diff
@@ -0,0 +1,147 @@
+
+    private void printJacORBEmulationWarningMessage() {
+        JacORBLogger.ROOT_LOGGER
+                .warn("Wildfly does not use JacORB as an ORB implementation anymore. JacORB subsystem would be emulated using "
```
INFO levels and up must be localized.

ctomc commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBSubsystemAdd.java:
```diff
@@ -0,0 +1,147 @@
+            throws OperationFailedException {
+        final ModelNode model = resource.getModel();
+        for (AttributeDefinition attr : JacORBSubsystemDefinitions.SUBSYSTEM_ATTRIBUTES) {
+            attr.validateAndSet(operation, model);
```
you could avoid this for loop by just calling super(JacORBSubsystemDefinitions.SUBSYSTEM_ATTRIBUTES) in constructor.

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBSubsystemDefinitions.java:
```diff
@@ -139,7 +140,6 @@
     // connection attribute definitions.
     public static final SimpleAttributeDefinition ORB_CONN_RETRIES = new SimpleAttributeDefinitionBuilder(
             JacORBSubsystemConstants.ORB_CONN_RETRIES, ModelType.INT, true)
-            .setDefaultValue(new ModelNode().set(5))
```
These default changes are a problem. It means a jacorb config that isn't used for EAP7 servers, but is used for EAP6 will now have different behavior. That's ok if the change is really necessary, but it's best avoided.

Actually, it's worse than that. Because the domain controller will say there's no default (or in the case of some other props, say that the default is "X") but then a legacy slave will see an undefined node and use "5", since it's code says "5". So if you change a default, you have to add a transformer such that if the param is undefined, you change the value to the new default.

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBSubsystemDefinitions.java:
```diff
@@ -492,11 +489,16 @@
+    // this it is necessary to modify legacy configuration.
+    static final SimpleAttributeDefinition PERSISTENT_SERVER_ID = new SimpleAttributeDefinitionBuilder(
+            Constants.ORB_PERSISTENT_SERVER_ID, ModelType.STRING, true).setDefaultValue(new ModelNode().set("1"))
+            .setFlags(AttributeAccess.Flag.RESTART_ALL_SERVICES).build();
```
This needs a transformer to reject it if defined when sending to a legacy slave.

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/logging/JacORBLogger.java:
```diff
@@ -445,4 +445,13 @@
+    OperationFailedException cannotEmulateProperty(String property);
+
+    @LogMessage(level = WARN)
+    @Message(id = 133, value = "Wildfly does not use JacORB as an ORB implementation anymore. JacORB subsystem would be "
```
Please don't use "WildFly" in messages, as it won't be WildFly once it's EAP. Use the passive voice to avoid the naming issue, i.e. "JacORB is not used as an ORB implementation any more."

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/logging/JacORBLogger.java:
```diff
@@ -445,4 +445,13 @@
+
+    @LogMessage(level = WARN)
+    @Message(id = 133, value = "Wildfly does not use JacORB as an ORB implementation anymore. JacORB subsystem would be "
+            + "emulated using OpenJDK ORB - implementation currently used by Wildfly. Ability to emulate legacy JacORB configurations "
```
"emulated using the current OpenJDK ORB implementation."

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBSubsystemAdd.java:
```diff
@@ -0,0 +1,146 @@
+                if (model.hasDefined(attribute.getName())) {
+                    throw JacORBLogger.ROOT_LOGGER.cannotEmulateProperty(attribute.getName());
+                }
+            }
```
I recommend creating a list of faulty attributes and then if not empty throwing the exception with the full list. Fixing them one at a time just to learn the next one would be annoying. :)

tadamski commented:
I have deprecated JacORB subsystem's root resource and all the properties. I have also refactored the subsystem and checked that integration tests+tck tests run when compatibility subsystem is used.

bstansberry commented:
Please don't deprecate the attributes. We aren't doing that for other resources where we are deprecating in EAP 6.4. We are just adding the deprecation metadata to the resources.

The issue is if a AttributeDefinition is configured as deprecated, and then validateAndSet is called on it, it writes an INFO message to the log. That is too noisy and we need to add some hook to better control it.

The log message is fine if the user can immediately take some action based on advice in the message, e.g. stop using attribute "foo" and switch to new attribute "bar". But if the user can't take action, IMO it's just spam in the logs.

In the case of this subsystem, we're supporting it for use on legacy slaves, and if it is used that way there is no action the user can take. The DC will end up with a lot of annoying logging.

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBExtension.java:
```diff
@@ -68,6 +70,8 @@
 
     private static final ModelVersion CURRENT_MODEL_VERSION = ModelVersion.create(1, 4, 0);
```
1.4.0 is the EAP 6.4 version. This one is different, and thus it needs to be 2.0.0.

You'll then need to transform to 1.4.0. To save duplication, consider using ChainedTransformationDescriptionBuilder, e.g. 

https://github.com/wildfly/wildfly-core/blob/master/jmx/src/main/java/org/jboss/as/jmx/JMXExtension.java#L174

That's not required though; duplicating is ok.

Also, transforming to a ModelVersion prior to the one in EAP 6.2.0 is not required, so you can drop any of that code if you like.

bstansberry replied:
EAP 6.2 used ModelVersion 1.3.0, so transforming to 1.2.x and earlier can be dropped.

Note that parsing previous xsds can't be dropped; that's a different thing.

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBSubsystemDefinitions.java:
```diff
@@ -402,101 +431,22 @@
+    // this it is necessary to modify legacy configuration.
+    static final SimpleAttributeDefinition PERSISTENT_SERVER_ID = new SimpleAttributeDefinitionBuilder(
+            Constants.ORB_PERSISTENT_SERVER_ID, ModelType.STRING, true).setDefaultValue(new ModelNode().set("1"))
+            .setFlags(AttributeAccess.Flag.RESTART_ALL_SERVICES).build();
```
You need a new xsd version 2.0 for this.

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBExtension.java:
```diff
@@ -185,11 +185,83 @@ protected static void registerTransformers_1_3(final SubsystemRegistration subsy
+        builder.getAttributeBuilder().setValueConverter(new ChangeUndefinedDefaultValueConverter(new ModelNode().set(5)),
+                JacORBSubsystemDefinitions.POA_REQUEST_PROC_POOL_SIZE);
+        builder.getAttributeBuilder().setValueConverter(new ChangeUndefinedDefaultValueConverter(new ModelNode().set(32)),
+                JacORBSubsystemDefinitions.POA_REQUEST_PROC_MAX_THREADS);
```
I apologize; I didn't do a cross-reference of the attributes where you changed the default value vs the ones JacORBSubsystemAdd is rejecting. If I had I would have realized what was going on and given different input.

These attributes are rejected on a current server if they are defined, but they still have their previous default values if they end up on a running server. Our attribute description metadata doesn't quite account for that concept, but I don't want to change the default value for this. I'd rather just have the text description of the attribute say that it will be rejected if configured on a running server.

This subsystem has 2 uses now -- legacy slave support on a DC, and a short term migration aid on current servers. Performing correctly for the legacy slave support case should be higher priority, as that's something users may use for many years. That means the default values should define what they truly are on the processes that support them, i.e. the legacy servers. The compatibility subsystem aspect should be to help make it easy to migrate, but it shouldn't redefine the attributes.

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBExtension.java:
```diff
@@ -185,11 +185,83 @@ protected static void registerTransformers_1_3(final SubsystemRegistration subsy
+        @Override
+        protected void convertAttribute(final PathAddress address, final String attributeName, final ModelNode attributeValue,
+                                        final TransformationContext context) {
+            if (!attributeValue.equals(newDefaultValue)) {
```
I don't think the "!" belongs here.

I don't think this conversion is right though. AIUI the attributeValue param should be what the user configured, likely undefined. If they specifically configured "NameService" I don't see why we would change it to something else. If they didn't configure it at all, that's where changing the value comes in, to make sure the legacy server config matches the current server config.

In this case though, is it correct to change the default value? Won't apps expect the JNDI binding to be what it used to be? Can this be handled in the OSH; i.e. convert undefined to "JBoss/Naming/root"? The iiop-jdk ORB should be able to handle that setting, right? Adding a ParameterCorrector to JacORBSubsystemDefinitions.NAMING_ROOT_CONTEXT can do the conversion of undefined to "JBoss/Naming/root".

bstansberry commented on legacy/jacorb/src/main/java/org/jboss/as/jacorb/JacORBSubsystemAdd.java:
```diff
@@ -0,0 +1,166 @@
+                if (model.hasDefined(attribute.getName())) {
+                    propertiesToReject.add(attribute.getName());
+                }
+            }
```
Similar processing needs to happen with the write-attribute op; i.e. if any of these attributes get defined on a server via write-attribute, the op needs to fail.

tadamski committed:
WFLY-3893 Deprecate JacORB subsystem

tadamski commented:
Fixed: Attributes are no longer deprecated. They have their default values which aren't used in HOST_CONTROLLER mode. Upgraded to 2.0.0 version. Added write attribute rejection. Passes all tck and interop tests.
